### PR TITLE
Enhance Panoptic Quality implementation

### LINKED ISSUE #1511: Extend Panoptic Quality usability, functionality and testing
## 🚀 Feature

Expand the [PanopticQuality](https://github.com/Lightning-AI/metrics/blob/master/src/torchmetrics/detection/panoptic_quality.py) code to
- allow batches in input (necessary feature)
- allow it to work on both images and point clouds (other inputs TBD)
- do not consider never seen classes in metric computation
- extend testing
- [let's see if something else comes up in the discussion here]

### Motivation

The just landed [PanopticQuality](https://github.com/Lightning-AI/metrics/blob/master/src/torchmetrics/detection/panoptic_quality.py) is a welcome addition to the toolbox, but there are a number of minor issues associated with it, from deal-breaking (does not take batches as inputs) to unnecessarily restrictive (PQ is a metric for panoptic segmentation both in point cloud and image domains), to suboptimal in terms of interpretability (if a class has TP + FP + FN = 0, why should it contribute to the final metric value?)

### Pitch

Let's bring in these goodies, happy to contribute 😉

### PR DESCRIPTION
**Pull request by marcocaccin**
## What does this PR do?

Fixes #1511:
- allow batches in input (necessary feature)
- allow inputs as images, point clouds, or data with any other spatial dimensionality
- do not consider never seen classes in metric computation
- make some progress in extending tests

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃
sure ✅

### CODE REVIEW

File: src/torchmetrics/detection/panoptic_quality.py [hunk 1]
```diff
@@ -48,11 +48,17 @@ class PanopticQuality(Metric):
             stuffs:
                 Set of ``category_id`` for uncountable stuffs.
             allow_unknown_preds_category:
-                Bool indication if unknown categories in preds is allowed
+                Boolean flag to specify if unknown categories in the predictions are to be ignored in the metric
+                computation or raise an exception when found.
+
+        TODO(marco): where to say that points in target that are not things or stuff
```
Comment (marcocaccin) [inline src/torchmetrics/detection/panoptic_quality.py:54]: This is something we should address in this PR, since it's not an obvious behaviour and just about every user needs to know how to define "ignore_index" kind of logic

File: src/torchmetrics/detection/panoptic_quality.py [hunk 2]
```diff
@@ -81,8 +87,8 @@ class PanopticQuality(Metric):
 
     def __init__(
         self,
-        things: Set[int],
-        stuffs: Set[int],
+        things: Collection[int],
```
Comment (marcocaccin) [inline src/torchmetrics/detection/panoptic_quality.py:90]: Made the input type generic, I don't see a reason to force the user to pass a `dict` - but we can discuss :)

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 3]
```diff
@@ -267,6 +283,51 @@ def _panoptic_quality_update(
...
+        - False negatives
+    """
+    device = flatten_preds.device
+    n_categories = len(cat_id_to_continuous_id)
+    iou_sum = torch.zeros(n_categories, dtype=torch.double, device=device)
+    true_positives = torch.zeros(n_categories, dtype=torch.int, device=device)
+    false_positives = torch.zeros(n_categories, dtype=torch.int, device=device)
+    false_negatives = torch.zeros(n_categories, dtype=torch.int, device=device)
+
+    # Loop over each sample independently: segments must not be matched across frames.
```
Comment (marcocaccin) [inline src/torchmetrics/functional/detection/panoptic_quality.py:315]: I wish there was a nicer vectorised implementation, but couldn't think of anything meaningful so far (was thinking about hashing the batch_idx inside the instance_id, but that's a hack)

File: tests/unittests/detection/test_panoptic_quality.py [hunk 4]
```diff
@@ -26,113 +27,143 @@
...
 
 
+def _compare_fn_0_1(preds, target) -> np.ndarray:
+    """Reference result for the _INPUTS_0, _ARGS_1 combination."""
+    return np.array([np.nan])
+
+
+def _compare_fn_1_2(preds, target) -> np.ndarray:
+    """Reference result for the _INPUTS_1, _ARGS_2 combination."""
+    return np.array([(2 / 3 + 1 + 2 / 3) / 3])
```
Comment (marcocaccin) [inline tests/unittests/detection/test_panoptic_quality.py:88]: These are kind of ugly, but lambda functions aren't allowed because of multiprocessing and a lookup table doesn't seem to make sense either when the signature of this function requires two tensors and only that

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 5]
```diff
@@ -282,22 +343,20 @@ def _panoptic_quality_compute(
...
     """
-    # TODO: exclude from mean categories that are never seen ?
-    # TODO: per class metrics
-
     # per category calculation
     denominator = (true_positives + 0.5 * false_positives + 0.5 * false_negatives).double()
     panoptic_quality = torch.where(denominator > 0.0, iou_sum / denominator, 0.0)
-    return torch.mean(panoptic_quality)
+    # Reduce across categories. TODO: is it useful to have the option of returning per class metrics?
+    return torch.mean(panoptic_quality[denominator > 0])
```
Comment (marcocaccin) [inline src/torchmetrics/functional/detection/panoptic_quality.py:352]: Note: this is a change in behaviour of the metric. Intuitively it does make sense to me, but we can argue whether there should be an option of masking out never seen classes or not

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/detection/panoptic_quality.py [hunk 6]
```diff
@@ -91,7 +98,7 @@ def __init__(
         # todo: better testing for correctness of metric
         warnings.warn("This is experimental version and are actively working on its stability.")
```
Comment (Borda) [inline src/torchmetrics/detection/panoptic_quality.py:99]:
```suggestion
```

File: src/torchmetrics/detection/panoptic_quality.py (hunk 1 above)
Comment (marcocaccin) [reply src/torchmetrics/detection/panoptic_quality.py]: [178e715](https://github.com/Lightning-AI/metrics/pull/1527/commits/178e7158cfa9a497d2ac2621075c4bdce5cde853)

File: src/torchmetrics/detection/panoptic_quality.py (hunk 2 above)
Comment (SkafteNicki) [reply src/torchmetrics/detection/panoptic_quality.py:93]: cool :)

File: src/torchmetrics/detection/panoptic_quality.py [hunk 7]
```diff
@@ -37,38 +37,47 @@ class PanopticQuality(Metric):
 
         where IOU, TP, FP and FN are respectively the sum of the intersection over union for true positives,
         the number of true postitives, false positives and false negatives. This metric is inspired by the PQ
-        implementati on of panopticapi, a standard implementation for the PQ metric for object detection.
+        implementation of panopticapi, a standard implementation for the PQ metric for panoptic segmentation.
 
     .. note:
-        Metric is currently experimental
+        Metric is currently experimental.
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/panoptic_quality.py:43]: this note can be removed now I would say :)

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 8]
```diff
@@ -33,55 +35,56 @@ def _to_tuple(t: Tensor) -> Tuple:
...
-        stuffs: All possible IDs for stuff categories
-    """
-    if not _is_set_int(things):
-        raise TypeError(f"Expected argument `things` to be of type `Set[int]`, but got {things}")
-    if not _is_set_int(stuffs):
-        raise TypeError(f"Expected argument `stuffs` to be of type `Set[int]`, but got {stuffs}")
-    if stuffs & things:
+    things_parsed = set(things)
+    if len(things_parsed) < len(things):
+        rank_zero_warn("The provided `things` categories contained duplicates, which have been removed.")
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/detection/panoptic_quality.py:73]:
```suggestion
        rank_zero_warn("The provided `things` categories contained duplicates, which have been removed.", UserWarning)
```

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 9]
```diff
@@ -33,55 +35,56 @@ def _to_tuple(t: Tensor) -> Tuple:
...
+    stuffs_parsed = set(stuffs)
+    if len(stuffs_parsed) < len(stuffs):
+        rank_zero_warn("The provided `stuffs` categories contained duplicates, which have been removed.")
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/detection/panoptic_quality.py:76]:
```suggestion
        rank_zero_warn("The provided `stuffs` categories contained duplicates, which have been removed.", UserWarning)
```

File: src/torchmetrics/functional/detection/panoptic_quality.py (hunk 3 above)
Comment (SkafteNicki) [reply src/torchmetrics/functional/detection/panoptic_quality.py:316]: yeah, it is not great to have for-loops inside the update method but I neither see an easy way of vectorizing this

File: src/torchmetrics/functional/detection/panoptic_quality.py (hunk 5 above)
Comment (SkafteNicki) [reply src/torchmetrics/functional/detection/panoptic_quality.py:353]: Fine by me :)

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 10]
```diff
@@ -309,6 +369,13 @@ def panoptic_quality(
     true postitives, false positives and false negatives. This metric is inspired by the PQ implementation of
     panopticapi, a standard implementation for the PQ metric for object detection.
 
+    .. note:
+        Metric is currently experimental.
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/detection/panoptic_quality.py:373]:
```suggestion
```

**Review by justusschock — APPROVED** [review]

File: src/torchmetrics/detection/panoptic_quality.py [hunk 11]
```diff
@@ -37,38 +37,47 @@ class PanopticQuality(Metric):
...
+
+    .. note:
```
Comment (Borda) [inline src/torchmetrics/detection/panoptic_quality.py:45]:
```suggestion
```

File: tests/unittests/detection/test_panoptic_quality.py (hunk 4 above)
Comment (Borda) [reply tests/unittests/detection/test_panoptic_quality.py:88]: that is fine :)

File: src/torchmetrics/functional/detection/panoptic_quality.py [hunk 12]
```diff
@@ -91,27 +94,35 @@ def _validate_inputs(preds: Tensor, target: torch.Tensor) -> None:
...
-    if not isinstance(target, torch.Tensor):
+    if not isinstance(target, Tensor):
         raise TypeError(f"Expected argument `target` to be of type `torch.Tensor`, but got {type(target)}")
     if preds.shape != target.shape:
         raise ValueError(
             f"Expected argument `preds` and `target` to have the same shape, but got {preds.shape} and {target.shape}"
         )
-    if preds.dim() != 3 or preds.shape[-1] != 2:
-        raise ValueError(f"Expected argument `preds` to have shape [height, width, 2], but got {preds.shape}")
+    if preds.dim() < 3:
```
[the remark this answers is not in the document]
Comment (Borda) [reply src/torchmetrics/functional/detection/panoptic_quality.py:105]: That is veru good question, so lets play with it in follow-up PR :)
cc: @Lightning-AI/core-metrics
